spec/orders.mediawiki

```diff
@@ -582,29 +584,65 @@ the checksum generated from their local copy of the epoch queue.
 ! field   !! size (bytes)  !! description
 |-
 | pimg    || string || hex-encoded preimage for the order's commitment
+|-
+| ackid   || int || the message ID to be used for the server's preimage acknowledgement
 |}
```

Comment:
I understand this is part of a solution to the need for a response to a response, but I'm not sure if/how it will work out allowing the client to specify the message id for an outgoing server message.  Although I think only requests really need to use `comms.NextID` to ensure a unique ID while it does not matter for notifications and responses.  Pretty sure this is not a problem with that?

An alternative would be do change the acknowledgement payload to contain an ackid also so that the containing message's ID would continue to be serial, monotonically increasing.

Also, part of the need for a response to a response was in the case of an error during response handling requiring the server to inform the client of the error in processing their response, so could we use this ackid concept for errors? Maybe new fields in the acknowledgement payload for success and general-purpose text details along with the signature?

Reply:
What do you think about flipping the semantics a little bit, and instead of sending a response to a response, we can instead send request for a request? For instance, the DEX-originating `preimage` request triggers the client to submit a `submit_preimage` request that just references either the `preimage` request ID, or maybe even just the market and order ID. It would require some reworking on the server side, since we can't use a request callback to process the preimage data. Just a thought.

Reply:
I think either request to a request, or a response to a response could be made to work.  I've been mulling over the protocol recently I was imagining a request or response with *no id set* to indicate that it's not a usual request/response and there's a reference id in the payload instead, and it should be treated as both a request (requires a response) and a response (there should be a registered response handler).

Although, defining a new request like `submit_preimage` would be perhaps cleaner if not as generalized to the myriad of cases where we need this kind of communication sequence such as the match ack response needing its own response (https://github.com/decred/dcrdex/pull/565/files#r460186427).

I guess one of us just needs to try out some ideas and start hacking.

Reply:
> Alt. the id could be set in the usual place, but a new msgjson.Dialog type that signals the dual role.

I like this variation. The payload can contain something like a sub-route as a sequencing field.